compliance/azure/ahub_manual/Azure_AHUB_Utilization_with_Manual_Entry.pt

```diff
@@ -0,0 +1,298 @@
+    client_id cred("AZURE_APPLICATION_ID")
+    client_secret cred("AZURE_APPLICATION_KEY")
+    additional_params do {
+      "resource" => "https://management.azure.com/"
+    } end
+  end
+end
+
+###############################################################################
+# Pagination
+###############################################################################
+
+pagination "azure_pagination" do
+  get_page_marker do
+    body_path "nextLink"
+  end
+  set_page_marker do
+    uri true
+  end
+end
+
+###############################################################################
+# Datasources
+###############################################################################
+
+#https://docs.microsoft.com/en-us/rest/api/compute/virtualmachines/listall
+datasource "ds_azure_instances" do
+  request do
+    auth $azure_auth
+    pagination $azure_pagination
+    host "management.azure.com"
+    path join(["/subscriptions/", $param_azure_sub, "/providers/Microsoft.Compute/virtualMachines"])
+    query "api-version","2018-06-01"
+    header "User-Agent", "RS Policies"
+  end
+  result do
+    encoding "json"
+    collect jmes_path(response, "value") do
+      field "id", jmes_path(col_item,"id")
+      field "name", jmes_path(col_item,"name")
+      field "location", jmes_path(col_item,"location")
+      field "licenseType", jmes_path(col_item,"properties.licenseType")
+      field "vmSize", jmes_path(col_item,"properties.hardwareProfile.vmSize")
+      field "osType", jmes_path(col_item,"properties.storageProfile.osDisk.osType")
+      field "imagePublisher", jmes_path(col_item,"properties.storageProfile.imageReference.publisher")
+      field "imageOffer", jmes_path(col_item,"properties.storageProfile.imageReference.offer")
+      field "imageSku", jmes_path(col_item,"properties.storageProfile.imageReference.sku")
+      field "imageVersion", jmes_path(col_item,"properties.storageProfile.imageReference.version")
+      field "tags", jmes_path(col_item,"tags")
+    end
+  end
+end
+
+datasource "ds_filtered_locations" do
+  run_script $js_location_list, $ds_azure_instances
+end
+
+#https://docs.microsoft.com/en-us/rest/api/compute/virtualmachinesizes/list
+datasource "ds_azure_vmsize" do
+iterate $ds_filtered_locations
+  request do
+    auth $azure_auth
+    host "management.azure.com"
+    path join(["/subscriptions/", $param_azure_sub, "/providers/Microsoft.Compute/locations/",val(iter_item, "location"),"/vmSizes"])
+    query "api-version","2018-06-01"
+    header "User-Agent", "RS Policies"
+  end
+  result do
+    encoding "json"
+    collect jmes_path(response, "value[*]") do
+      field "virtualmachineSize", jmes_path(col_item,"name")
+      field "numberOfCores", jmes_path(col_item,"numberOfCores")
+    end
+  end
+end
+
+datasource "ds_vminstance_with_cores" do
+  run_script $js_vminstances_with_cores, $ds_azure_vmsize, $ds_azure_instances, $param_exclusion_tag_key, $param_license_count
+end
+
+
+datasource "ds_results_size" do
```

Comment:
I don't think this is necessary.  It appears you are only using this for the check.  in the check you can still use final_result.result.  You should be able use something like *check eq(size(val(data, "result")), 0) *
See example. https://github.com/flexera/policy_templates/blob/master/cost/downsize_instance/downsize_instance.pt#L332
You can remove ds_results_size and js_results_size.

Reply:
Thank you Curt, we were not aware of this . Now I have taken care of this and tested the same.